ci: test against Juju 4.0.0

tonyandrewmeyer opened the pull request:
This PR adds testing against Juju 4, and compatibility so that the integration tests pass.

There are a few issues blocking the tests passing where we don't want to make changes in Jubilant:
 * [x] [`show-secret name` shows all secrets](https://github.com/juju/juju/issues/21296)
 * [ ] [Cross-model integration doesn't trigger all events](https://github.com/juju/juju/issues/21412)
 * [ ] [App status isn't derived from unit status](https://github.com/juju/juju/issues/21424)

Some small drive-bys:
 * Change from testing against MicroK8s to Canonical K8s
 * Change from testing against Python 3.13 to Python 3.14
 * Fix `make help` to include `integration-k8s` and any other future targets that contain digits
 * Expose the `expires` field in a secret revision, in both Juju 3 and Juju 4 (I suspect this was missed because the secret classes weren't autogenerated from the Go structs, and it's only included if there is an expiry)

[A run in my fork](https://github.com/tonyandrewmeyer/jubilant/actions/runs/20251015399).

Fixes #229

tonyandrewmeyer committed:
Try testing against 4/beta.

tonyandrewmeyer committed:
Switch from beta to candidate.

tonyandrewmeyer committed:
Use a proper matrix.

tonyandrewmeyer committed:
Add missing $

tonyandrewmeyer committed:
Attempt to fix k8s.

tonyandrewmeyer committed:
Typo

tonyandrewmeyer committed:
Temporarily add additionalProperties to work around the Juju bug.

tonyandrewmeyer committed:
If an action is not defined in Juju 4.0, continue to raise a ValueError.

tonyandrewmeyer committed:
Work around Juju 4 bug for now.

tonyandrewmeyer committed:
chore: format

tonyandrewmeyer committed:
Look for the message in the unit workload status, not app status.

tonyandrewmeyer committed:
Wrap comment, provide initial values to make pyright happy.

tonyandrewmeyer committed:
Add/remove SSH keys to test that functionality and also make ssh and scp work on Juju 4.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Fix secret workaround.

tonyandrewmeyer committed:
Handle Juju as a snap.

tonyandrewmeyer committed:
Add/remove SSH key for the machine test.

tonyandrewmeyer committed:
Add missing imports.

tonyandrewmeyer committed:
Ensure that ssh-keygen is available.

tonyandrewmeyer committed:
Simplify by using a hard-coded test key.

tonyandrewmeyer committed:
No SSH key management for K8s tests.

tonyandrewmeyer committed:
More workarounds for the Juju bug.

tonyandrewmeyer committed:
Typo

tonyandrewmeyer committed:
Juju 4 does not provide a backend field in the revision struct.

tonyandrewmeyer committed:
Get the message from the unit, to be compatible with both 3 and 4.

tonyandrewmeyer committed:
Add a sleep to work around the SSH key not being available immediately, and change the SSH key.

tonyandrewmeyer committed:
Try waiting a crazy amount of time.

tonyandrewmeyer committed:
Fix unit name.

tonyandrewmeyer committed:
Make linter happy.

tonyandrewmeyer committed:
Try a longer timeout.

tonyandrewmeyer committed:
Merge remote-tracking branch 'origin/main' into juju4-integration

tonyandrewmeyer commented on jubilant/_juju.py:
```diff
@@ -1057,7 +1057,14 @@ def run(
+                if (
+                    'task failed' not in exc.stderr
+                    and 'inserting operation action' not in exc.stderr
+                ):
```
Juju 3 exits with a zero code, Juju 4 with a non-zero code. I think the Juju 4 behaviour is more correct so Jubilant should handle both.

We could perhaps suggest improvements to the error message that would make it easier for humans to understand and also allow a nicer matching string.

benhoyt replied:
Per discussion, let's check that the exception is consistent between Juju 3 and 4. It looks like it'll fall down to the raise ValueError "error running action" for Juju 3 (if return code is zero). But we should check and be consistent.

tonyandrewmeyer replied:
https://github.com/juju/juju/issues/21446

tonyandrewmeyer commented on tests/integration/charms/testdb/charmcraft.yaml:
```diff
@@ -23,6 +23,7 @@ actions:
     params:
       mode:
         type: string
+    additionalProperties: true
```
Apparently swapping the default value here is a deliberate breaking change, albeit one that is not yet documented. Being explicit seems a reasonable thing to do anyway.

See: https://github.com/juju/juju/issues/21294

tonyandrewmeyer commented on tests/integration/helpers.py:
```diff
@@ -4,6 +4,18 @@
+"""
+
+TEST_SSH_PUBLIC_KEY = 'ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIMECk+zN/qSL1U0BakCxTpVjWIzn2p7kfITTzOvy3DRK tameyer@tam-canoncial-1'
+
```
I don't love having a key in the test code (and lots of thing yell at you about it) even though it doesn't have any value at all. However, `ssh-keygen` isn't available in the GitHub environment, and going through the steps to install it was quite involved. I'm happy to hear suggestions about alternatives (I do think we want a separate one, not to us the user's existing key, for more explicit testing, both in 3 and 4).

benhoyt replied:
Let's see if we can use the `openssl` tool for this (eg: `openssl genrsa`).

tonyandrewmeyer commented on tests/integration/test_cross_model.py:
```diff
@@ -14,9 +14,12 @@ def test_offer_and_consume(juju: jubilant.Juju, model2: jubilant.Juju):
+    assert status.apps['testdb'].units['testdb/0'].workload_status.message == 'relation created'
 
-    status2 = model2.wait(jubilant.all_active)
+    status2 = model2.wait(jubilant.all_active, timeout=900)
```
Juju 4 just seems reliably slower in this case. 900 is a guess that seems to work.

I'm not sure whether opening a Juju ticket for a performance regression like this makes sense.

tonyandrewmeyer commented on tests/integration/test_machine.py:
```diff
@@ -25,11 +47,20 @@ def test_exec(juju: jubilant.Juju):
-    output = juju.ssh('ubuntu/0', 'echo', 'UNIT')
+def test_ssh(juju: jubilant.Juju, private_key_file: str):
+    # The key is not available for use immediately. For now, just wait for a
+    # moment. Waiting for `ssh-keys` to not be empty does not work as a solution.
```
I think the best choice here would be to adjust the test order so that the ssh key is added early on (maybe an explicit test not a fixture) so by the time we get to this test it's in place.

benhoyt replied:
I actually think being explicit here is best. Your retry loop or something like it seems better than relying an earlier test happening some time ago.

benhoyt commented on jubilant/_juju.py:
```diff
@@ -1219,9 +1226,27 @@ def show_secret(
+            if uri_from_juju == identifier or ('name' in obj and obj['name'] == identifier):
+                break
+            # Allow falling through, which will give the first secret,
+            # which is correct in Juju 3.
```
I agree. Are these two separate issues? In any case, I agree -- I don't think we should work around either of them in Jubilant. They seem like straight Juju bugs which are fairly high priority if using secrets.

benhoyt commented on tests/integration/test_cross_model.py:
```diff
@@ -14,9 +14,12 @@ def test_offer_and_consume(juju: jubilant.Juju, model2: jubilant.Juju):
 
     status = juju.wait(jubilant.all_active)
-    assert status.apps['testdb'].app_status.message == 'relation created'
+    assert status.apps['testdb'].units['testdb/0'].workload_status.message == 'relation created'
```
I agree it seems reasonable to change the test in this case. But thanks for investigating and opening the bug -- I think that will affect a lot of charms and hence is pretty high pri.

benhoyt commented on tests/integration/test_machine.py:
```diff
@@ -25,11 +47,20 @@ def test_exec(juju: jubilant.Juju):
+    output = None
+    for _ in range(60):
+        try:
+            output = juju.ssh('ubuntu/0', 'echo', 'UNIT', ssh_options=['-i', private_key_file])
```
Shouldn't this break out of the loop if it's good?

benhoyt reviewed (commented):
Thanks for this. I've done an initial review and left a few comments.

tonyandrewmeyer committed:
Drop testing against MicroK8s.

tonyandrewmeyer committed:
Drop the workaround for the show-secret Juju bug -- we'll get charmers to push Juju for a fix instead.

tonyandrewmeyer commented on jubilant/secrettypes.py:
```diff
@@ -103,7 +103,7 @@ class Revision:
         return cls(
             revision=d['revision'],
-            backend=d['backend'],
+            backend=d.get('backend', ''),
```
Opened https://github.com/juju/juju/issues/21448. However, the struct has `omitempty` for `backend` and `expiretime` so we should handle those being missing.

tonyandrewmeyer committed:
Drive-by: add missing revision field.

tonyandrewmeyer committed:
Go back to looking for the message on the app - we'll wait for Juju to fix the issue where it doesn't update.

tonyandrewmeyer committed:
Add missing break. Only retry for a minute, not 10 minutes.

tonyandrewmeyer committed:
Go back to looking for the message on the app - we'll wait for Juju to fix the issue where it doesn't update.

tonyandrewmeyer committed:
Generate a SSH key on the fly so that we don't have one in the source repository.

openssl by itself as an executable can't get us all the way, as far as I can tell, but the cryptography package can do everything, so use that (and avoid openssl as a binary). This does introduce a new dependency, but it's only for testing, so it seems ok, I think.

tonyandrewmeyer committed:
help should display commands that have digits in them too, like integration-k8s (drive-by)

tonyandrewmeyer committed:
It seems like the longer timeout is not required.

tonyandrewmeyer committed:
Fix import order.

benhoyt reviewed (changes requested):
Couple of questions / changes requested.

benhoyt commented on jubilant/secrettypes.py:
```diff
@@ -98,14 +98,16 @@ class Revision:
     backend: str
     created: datetime.datetime
     updated: datetime.datetime
+    expires: datetime.datetime | None
```
Oops, my bad -- didn't realise the field was new.

tonyandrewmeyer committed:
Update tests/integration/helpers.py

tonyandrewmeyer commented on tests/integration/helpers.py:
```diff
@@ -2,9 +2,39 @@
+    public_key_ssh = public_key.public_bytes(
+        encoding=cryptography.hazmat.primitives.serialization.Encoding.OpenSSH,
+        format=cryptography.hazmat.primitives.serialization.PublicFormat.OpenSSH,
+    ).decode('utf-8')
```
```suggestion
    ).decode()
```

tonyandrewmeyer committed:
Update tests/integration/helpers.py

tonyandrewmeyer committed:
Exclude the run() fix from this PR, per discussion.

tonyandrewmeyer committed:
Adjust imports.

benhoyt commented:
I reckon just keep them as a matrix and update the [repo automation](https://github.com/canonical/canonical-repo-automation/blob/main/groups/charm-engineering/charm-tech/repos/jubilant/terragrunt.hcl) to require `integration-k8s (3/stable)` and `integration-machine (3/stable)` specifically.

tonyandrewmeyer committed:
Merge branch 'main' into juju4-integration
